From d3a3855b1ddbd3fa445e463d23082abda2e0d921 Mon Sep 17 00:00:00 2001
From: anonym <anonym@riseup.net>
Date: Wed, 27 Feb 2019 16:15:59 +0100
Subject: [PATCH] Make use of non-SSL Exchange AutoDiscover methods optional.

If an attacker does a MitM they can presumably modify the Exchange
server's HTTP response to redirect to an attacker controller Exchange
server instead. So let's provide protection against this via the
mailnews.auto_config.sslOnly pref.
---
  .../content/exchangeAutoDiscover.js           | 20 ++++++++++---------
 1 file changed, 11 insertions(+), 9 deletions(-)

--- a/comm/mail/components/accountcreation/content/exchangeAutoDiscover.js
+++ b/comm/mail/components/accountcreation/content/exchangeAutoDiscover.js
@@ -127,15 +127,17 @@
   fetch.start();
   call.setAbortable(fetch);

-  call = priority.addCall();
-  fetch3 = new FetchHTTP(
-    url3,
-    callArgs,
-    call.successCallback(),
-    call.errorCallback()
-  );
-  fetch3.start();
-  call.setAbortable(fetch3);
+  if (!Services.prefs.getBoolPref("mailnews.auto_config.sslOnly")) {
+    call = priority.addCall();
+    fetch3 = new FetchHTTP(
+      url3,
+      callArgs,
+      call.successCallback(),
+      call.errorCallback()
+    );
+    fetch3.start();
+    call.setAbortable(fetch3);
+  }

   // url3 is an HTTP URL that will redirect to the real one, usually a HTTPS
   // URL of the hoster. XMLHttpRequest unfortunately loses the call
